ShellDriver: implement file transfer support via XMODEM

rohieb opened the pull request:
This implementation starts the respective sender or receiver on the target and then uses tehmaze's [xmodem][] Python implementation to transfer files over the serial console. It can work with busybox's `rx` (only for receiving on the target) or `lrz`/`lsz` from [lrzsz][] on the target, and autodetects which commands are available.

XMODEM is a very simple, but also very historic protocol with some quirks. See the comment in `labgrid.driver.ShellDriver._put()` for some insight and workarounds.

Unfortunately, after the peer on the target has been started, there is no way to stop it if the transfer fails and get back to a shell prompt, since the peer on the target interprets everything on the console as XMODEM messages. lsz's `--min-bps` option does not seem to work very good (but are included nevertheless), so powercycling the target is possibly the best way to get rid of such situations.

[xmodem]: https://github.com/tehmaze/xmodem
[lrzsz]: https://ohse.de/uwe/software/lrzsz.html

jluebbe commented on labgrid/driver/shelldriver.py:
```diff
@@ -203,3 +205,127 @@ def _put_ssh_key(self, key):
+        elif self._run("which lrz")[0]:
+            # redirect stderr to prevent lrz from printing "ready to receive
+            # $file", which will confuse the XMODEM instance
+            cmd = "lrz -X -y -c --o-sync '{}' 2>/dev/null".format(tmpfile)
```
We shouldn't need -o-sync. Do we need -b?

jluebbe commented on labgrid/driver/shelldriver.py:
```diff
@@ -203,3 +205,127 @@ def _put_ssh_key(self, key):
+            # cmd now", which will confuse the XMODEM instance
+            cmd = "lsz -b -X -m 1200 -M 10 '{}' 2>/dev/null".format(remotefile)
+        else:
+            raise ExecutionError("No XMODEM sender (lsz) available on target")
```
Also do caching as above.

rohieb commented:
These commits should address your feedback, feel free to sqash them, or ask for more ;)

jluebbe commented on labgrid/driver/shelldriver.py:
```diff
@@ -274,23 +274,31 @@ def _put(self, localfile, remotefile):
         with open(localfile, "rb") as ifh:
+
+            # We first write to a temp file, which we'll `dd` onto the destination file later
+            tmpfile = self._run("mktemp")[0]
```
Check the return value or use _run_check.

rohieb replied:
Right, done.

jluebbe commented:
Please also squash the changes into a single commit.

rohieb commented on labgrid/driver/shelldriver.py:
```diff
@@ -203,3 +208,147 @@ def _put_ssh_key(self, key):
+
+        _target_cleanup(tmpfile)
+        if ret != 0:
+            raise ExecutionError("Could not truncate destination file: dd returned %d", ret)
```
Oops, that should have been a `"%d" % ret`. Fixed.

rohieb commented:
Forgot to add a Signed-off-by line.

rohieb commented:
I've now also included a way to upload and run scripts on the target, as a convenience for the user; and extended the documentation a bit.

rohieb commented:
OK, the feature set should be complete by now. I've added functionality to upload arbitrary data from byte arrays instead of files, and also rebased to current master.
